dump-acct.8
.\" Copyright (C) 2009 Red Hat, Inc. All Rights Reserved.
.\" Written by Ivana Hutarova Varekova <varekova@redhat.com>
.TH DUMP_ACCT 8 "2009 December 2" "Linux accounting system"

.SH NAME
dump\-acct \- print an acct/pacct file in human-readable format

.SH SYNOPSIS
.B dump-acct
[\fB\-r\fR|\fB\-\-reverse\fR]
[\fB\-R\fR|\fB\-\-raw\fR]
[\fB\-n\fR|\fB\-\-num\fR\ \fIrecs\fR]
[\fB\-\-byte\-swap\fR]
[\fB\-\-format\fR]
[\fB\-\-ahz\fR\ \fIfreq\fR]
[\fB\-h\fR|\fB\-\-help\fR]
[\fIfiles\fR]

.SH DESCRIPTION
The dump-acct command transforms the output \fBfile\fR from the
accton format to the human-readable format: one record per line.
Each record consists of severald fields which are separated by
character "|" (the meaning of concreate field depends on the
version of kernel package - with which the accton file was created).

.SH OPTIONS
The following options are supported:
.TP
.B -r, --reverse
Print the output in reverse order.
.TP
.B -R, --raw
The records will be printed without any parsing.
.TP
.B -n, --num NUMRECS
Display only the first NUMRECS number of records.
.TP
.B --byteswap
Swap the bytes (relative to your system's native byte order) in --raw output.
.TP
.B --format
Set output format with --raw option.
.TP
.B --ahz FREQ
Set the AHZ (platform dependent frequency in Hertz) to FREQ.
.TP
.B -h, --help
Print a help message and the default location of the process accounting file
and exit.

.SH SEE ALSO
accton (8), lastcomm (1)



NEWS
New in 6.6.3:

 * Fixed manuals.

New in 6.6.2:

  * Link with -lm.
  * Fix texi.
  * The rest of fixes from 6.5.5 to 6.6.1.  

New in 6.5.4:

  * Fix for BSDen.

New in 6.5.3:

  * Merged patches from Debian and Red Hat.

New in 6.5.1:

  * GNUlib for compability, audited code... Reworked.

New in 6.5:

  * Patches from Debian.

  * Actually released to the public.

  * Huge code-refactoring.

New in 6.4:

  * Linux multiformat support: Starting from release 2.6.8, the Linux
    kernel writes either the totally new v3 format or v1 or v2 formats
    that are enhanced relative to the v0 format of earlier kernels.
    GNU acct now understands v0, v2, and v3 formats (v1, written on
    the m68k architecture, is not supported) and uses the information
    from the additional fields in v2 format. Particularely, 32 bit
    user and group IDs are supported.

  * Added `--format', `--byteswap', and `--ahz' options to dump-acct.
    Together with the existing `--raw' switch and Linux multiformat
    support this results in a handy format converter.

  * The acct.h header file is now expected to be at <sys/acct.h>,
    the older <linux/acct.h> is ignored. As this file is not used
    at all when Linux multiformat is enabled (which is the default),
    noone should notice this change anyways.

  * A bug with the ctime() function on 64 bit platforms was fixed.

  * Some small code cleanups were performed, GNU acct should now compile
    without warnings on ANSI C compilers.

  * The build system was updated to recent versions of the tools used.


New in 6.3.5:
  * Reverted below change to acct.h header search.


New in 6.3.4:
  * When trying to find the acct.h header file, look in the linux
    directory before the sys directory, since the former is probably
    newer than the latter.

  * Documentation fixes for sa.


New in 6.3.3:

  * For those systems which have wide utmp fields, last now tries to
    fit its output into 80 columns by using "reasonable" default
    fields lengths for those members.  The `-w' and `--wide' flags
    have been added to allow you to view the fields with their "full
    precision," as it were.

  * Fix elapsed time calculation in sa.

  * Added an `--ahz' flag to lastcomm and sa to make it easier for
    those folks who are trying to analyze logs generated on other
    systems with the same byte order and file format but a different
    value for AHZ.

  * Fixed the `--strict-match' flag in lastcomm.  The following now
    works as expected:

      lastcomm --strict-match noel cat ttyp7


New in 6.3.2:

  * Small memory allocation problem fixed in sa.

  * Released to the general public.


New in 6.3.1:

  * A `NEWS' file (and others) so that we conform to the GNU
    requirements for packaging.

  * We use `automake' to build the package, so things like ansi2knr
    should be called automatically for those of you that don't have
    ANSI C compilers.  It also helps write the install targets, etc.
    Very cool.  I was suprised how easy it was to integrate into this
    package.  Hats off to David J. MacKenzie and Tom Tromey!


New in 6.3:

  * Hey!  We're on prep.ai.mit.edu!

  * Renamed long option `--year' to `--print-year' in both the ac and
  last programs.

  * Added a `--print-seconds' (or `-s') flag to last, for those who
  really want the nitty-gritty.

  * Added a `--sort-real-time' option to sa for those who want to sort
  the output by "real" (elapsed) time.

  * Fixed a buffer overrun error in sa which was preventing it from
  sorting the username list correctly.

Ancient history:

  * Check the file `ChangeLog' in the source directory.


dump-acct.c
/*
Copyright (C) 1997, 2003, 2005, 2008, 2009, 2010 Free Software Foundation, Inc.

This file is part of the GNU Accounting Utilities

The GNU Accounting Utilities are free software; you can redistribute
them and/or modify them under the terms of the GNU General Public
License as published by the Free Software Foundation; either version
3, or (at your option) any later version.

The GNU Accounting Utilities are distributed in the hope that they will
be useful, but WITHOUT ANY WARRANTY; without even the implied warranty
of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License
along with the GNU Accounting Utilities; see the file COPYING.  If
not, write to the Free Software Foundation, 675 Mass Ave, Cambridge,
MA 02139, USA.
*/

/* print an acct/pacct file in human-readable format. */

#include "config.h"

#include <stdio.h>

#ifdef HAVE_STDLIB_H
#include <stdlib.h>
#endif

#include "common.h"
#include "files.h"
#include "pacct_rd.h"
#ifdef HAVE_GETOPT_LONG_ONLY
#include <getopt.h>
#else
#include "getopt.h"
#endif

char *program_name;
int debugging_enabled = 0;	/* no -- we don't care about bad
				   records or the file-reading
				   algorithms. */
int ahz = AHZ;

void give_usage(void)
{
  printf ("Usage: %s [-hrR] [-n <recs>] <files>\n\
          [--num <recs>] [--raw] [--reverse] [--help]\n",
          program_name);
#ifdef LINUX_MULTIFORMAT
  printf ("          [--format <version>] [--byteswap]\n");
#endif
  printf ("          [--ahz <freq>]\n");
  print_acct_file_location ();
}

int main(int argc, char *argv[])
{
  int c;
  int backwards_flag = 0;	/* start forwards */
  long num_lines_to_print = -1;	/* -1 means all */
  int raw = 0;
#ifdef LINUX_MULTIFORMAT
  int version = -1;
  int byteswap = 0;
  size_t size;
  static struct acct buf;
#endif

  program_name = argv[0];

  while (1)
    {
      int option_index = 0;

      static struct option long_options[] =
      {
        { "reverse", no_argument, NULL, 1
        },
        { "help", no_argument, NULL, 2 },
        { "num", required_argument, NULL, 3 },
        { "raw", no_argument, NULL, 4 },
        { "ahz", required_argument, NULL, 5 },
#ifdef LINUX_MULTIFORMAT
        { "format", required_argument, NULL, 6 },
        { "byteswap", no_argument, NULL, 7 },
#endif
        { 0, 0, 0, 0 },
      };

      c = getopt_long (argc, argv, "rhn:R", long_options, &option_index);

      if (c == EOF)
        break;

      switch (c)
        {
        case 'r':
        case 1:
          backwards_flag = 1;
          break;
        case 'h':
        case 2:
          give_usage();
          exit (EXIT_FAILURE);
        case 'n':
        case 3:
          num_lines_to_print = strtol(optarg, (char **)NULL, 10);
          if (num_lines_to_print < 1)
            fatal ("number of lines to print must be positive and non-zero");
          break;
        case 'R':
        case 4:
          raw = 1;
          break;
        case 5:
          ahz = strtol(optarg, (char **)NULL, 10);
          if (ahz < 1)
            {
              (void)printf ("%s: AHZ must be one or greater\n", program_name);
              exit (EXIT_FAILURE);
            }
          break;
#ifdef LINUX_MULTIFORMAT
        case 6:
          if (optarg[0]=='v')
            optarg++;
          version = strtol(optarg, (char **)NULL, 10);
          break;
        case 7:
          byteswap = 1;
          break;
#endif
        }
    }

  if (optind >= argc)
    {
      /* User didn't pass any filenames -- give them the usage
         message. */
      give_usage ();
      exit (EXIT_FAILURE);
    }

  pacct_init (backwards_flag);

  /* Add all files that we want to dump. */

  while (optind < argc)
    add_pacct_file (argv[optind++]);

  /* Dump the records. */

  {
    struct acct *rec;
    while ((rec = pacct_get_entry ()) != NULL)
      {
        if (raw)
          {
#ifdef LINUX_MULTIFORMAT
            size = convert_acct_record_write (rec, &buf, version, byteswap);
            (void)fwrite (&buf, size, 1, stdout);
#else
            (void)fwrite (rec, sizeof (struct acct), 1, stdout);
#endif
          }
        else
          print_pacct_record (rec, stdout);

        if (num_lines_to_print > -1)
          {
            num_lines_to_print--;
            if (num_lines_to_print == 0)    /* max lines printed */
              exit (EXIT_SUCCESS);
          }
      }

  }
  exit (EXIT_SUCCESS);
}
